### PR TITLE
Implement project name in dock tile

### PR DESCRIPTION
**Pull request by timtmok**
### Intent
Address #11386

### Approach
Port the native code that redraws the application icon with the project name overlaid. Add in stubs for the other platforms.

I think we can implement this for Windows and Linux since there's support to set an icon: https://www.electronjs.org/docs/latest/api/browser-window#winseticonicon-windows-linux

It wouldn't need native code either! (still needs it for Mac since that API is supported for Windows and Linux only)

image

I've also reorganized the native code folder in case we add more modules. The type definitions go in the root and the actual code are in their respective folders.

### Automated Tests
None

### QA Notes
This is the same code as Qt so the look should be exactly the same.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

File: src/node/desktop/src/main/gwt-callback.ts [hunk 1]
```diff
@@ -544,6 +544,7 @@ export class GwtCallback extends EventEmitter {
 
     ipcMain.on('desktop_bring_main_frame_to_front', () => {
       this.mainWindow.window.focus();
+      console.debug('desktop_bring_main_frame_to_front');
```
Comment (gtritchie) [inline src/node/desktop/src/main/gwt-callback.ts:547]: Did you mean to leave this in?

File: src/node/desktop/src/native/dock/dock_mac.mm [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+#define RS_EXPORT_FUNCTION(__NAME__, __FUNCTION__) \
+  exports.Set(                                     \
+    Napi::String::New(env, __NAME__),              \
+    Napi::Function::New(env, __FUNCTION__)         \
+  )
+
+// Implementation based on https://github.com/jessesquires/JSCustomBadge, and
+// https://cgit.kde.org/kdevplatform.git/commit/?id=c0626300e1c18448dd653a3484fd1ca6298e2d8c
+
+@interface MacDockTileView : NSView {
```
Comment (gtritchie) [inline src/node/desktop/src/native/dock/dock_mac.mm:32]: Ah, I have a soft spot for Objective-C. Was doing NeXTSTEP programming in it long, long ago.

**Review by gtritchie — APPROVED** [review]
LGTM! One comment about what looks like debugging code.

File: src/node/desktop/src/main/gwt-callback.ts (hunk 1 above)
Comment (timtmok) [reply src/node/desktop/src/main/gwt-callback.ts:547]: I'll clean this up. I got side tracked into trying to reproduce another bug!

**Pull request merged**